### PR TITLE
osrm-routed monitoring in the prometheus format

### PR DESCRIPTION
**Pull request by gardster**
# Issue

[#5180](https://github.com/Project-OSRM/osrm-backend/issues/5180)
Endpoint to check OSRM status with monitoring solutions based on [prometheus](https://prometheus.io/docs/concepts/data_model/) format. 
This is a proof of concept implementation that has only several simple options. Additional metrics can be added with ongoing PRs.
Metrics endpoint is a separate thread that listens its own port and doesn't bother worker threads.

### CODE REVIEW

File: features/step_definitions/monitoring.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+                    }
+
+
+                    if (!metrics.hasOwnProperty("osrm_routed_instance_info")) {
+                        throw new Error("Have no instance information inside the monitoring!");
+                    }
+                    cb(null, got);
+                };
+                this.requestMonitoring(afterRequest);
+            }
```
Comment (IharSha) [inline features/step_definitions/monitoring.js]:
```suggestion
            };
```

File: features/step_definitions/monitoring.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+                var afterRequest = (err, res) => {
+                    if (err) return cb(err);
+
+                    var rows = res.body.split('\n');
+                    var metrics = {};
+                    rows.forEach((r) => {
+                        if (r.includes('{')) {
+                            var k = r.split('{')[0],
+                                v = r.split('}')[1].split(' ')[1];
+                            metrics[k] = v;
```
Comment (IharSha) [inline features/step_definitions/monitoring.js]:
```suggestion
                            metrics[k] = r.split('}')[1].split(' ')[1];
```

File: features/step_definitions/monitoring.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            var testRow = (row, ri, cb) => {
+                var afterRequest = (err, res) => {
+                    if (err) return cb(err);
+
+                    var rows = res.body.split('\n');
+                    var metrics = {};
+                    rows.forEach((r) => {
+                        if (r.includes('{')) {
+                            var k = r.split('{')[0],
+                                v = r.split('}')[1].split(' ')[1];
```
Comment (IharSha) [inline features/step_definitions/monitoring.js]:
```suggestion
```

File: docs/monitoring.md [hunk 4]
```diff
@@ -0,0 +1,15 @@
+## Abstract
+
+OSRM routed daemon publish helath information in prometheus format.
```
Comment (IharSha) [inline docs/monitoring.md]:
```suggestion
OSRM routed daemon publish health information in prometheus format.
```

File: features/support/route.js [hunk 5]
```diff
@@ -16,6 +16,24 @@ module.exports = function () {
...
+            limit = Timeout(this.TIMEOUT, { err: { statusCode: 408 } });
+        function runRequest (cb) {
+            request(uri, cb);
+        }
+
+        runRequest(limit((err, res, body) => {
+            if (err) {
+                if (err.statusCode === 408) throw new Error('*** osrm monitoring endpoint did not respond');
+                else if (err.code === 'ECONNREFUSED')
+                    throw new Error('*** osrm monitoring endpoint is not running', uri);
```
Comment (IharSha) [inline features/support/route.js]: I'm not sure that `Error` can handle more than one arg.

File: features/step_definitions/monitoring.js (hunk 1 above)
Comment (gardster) [reply features/step_definitions/monitoring.js]: fixed

**Comment by gardster** [comment]
@IharSha thank you for the review. Fixed!

File: docs/monitoring.md [hunk 6]
```diff
@@ -0,0 +1,15 @@
+## Abstract
+
+OSRM routed daemon publish health information in prometheus format.
+This option switched off by default. To enable this feature please set `-P <monitoring_port>` option on OSRM routed startup.
+
+
+## Available metrics
+
+- `osrm_routed_instance_info` metric with base instanse information inside labels
```
Comment (ChristianAlexander) [inline docs/monitoring.md]: nit: instanse -> instance

Comment (gardster) [reply docs/monitoring.md]: Thank you! Fixed.

**Comment by gardster** [comment]
+ need to provide metrics behind the `/metrics` endpoint

File: docs/monitoring.md [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+
+## Available metrics
+
+- `osrm_routed_instance_info` metric with base instance information inside labels
+-- `algorithm` routing algorithm;
+-- `code_version` OSRM version. Same with `-v` command line option;
+-- `data_version` OSM data version if available. This value can be setted at the `osrm-extract` phase;
+-- `working_threads` actual count of http requests processing threads;
+- `http_requests_count` http calls count divided by plugins. And `invalid` count for requests were plugin wasn't set;
```
[the remark this answers is not in the document]
Comment (gardster) [reply docs/monitoring.md]: fixed

**Comment by atmosx** [comment]
Any idea when/if is this going to me merged?

**Comment by edumucelli** [comment]
It would be really nice to have this. We are heavily using OSRM and that is definitely something missing. We rely on the service mesh (Istio) dashboard to have metrics.

**Comment by cemo** [comment]
@gardster Why did not you merge this PR? Is there a reason for it?

**Comment by gardster** [comment]
@cemo I haven't rights to merge self PRs

**Comment by gardster** [comment]
@akashihi PTAL

File: CHANGELOG.md [hunk 8]
```diff
@@ -9,6 +9,7 @@
       - ADDED: keepalive support to the osrm-routed HTTP server [#5518](https://github.com/Project-OSRM/osrm-backend/pull/5518)
       - ADDED: flatbuffers output format support [#5513](https://github.com/Project-OSRM/osrm-backend/pull/5513)
       - ADDED: Global 'skip_waypoints' option [#5556](https://github.com/Project-OSRM/osrm-backend/pull/5556)
+      - ADDED: monitoring endpoint in prometheus format [#5180](https://github.com/Project-OSRM/osrm-backend/pull/5180)
```
Comment (akashihi) [inline CHANGELOG.md:12]: I guess this one needs to be at 5.24.0 section:)

**Review by akashihi — APPROVED** [review]

**Comment by wangyoucao577** [comment]
Why use a separate listen port instead of same port of other services?

**Comment by ugurakkurt** [comment]
We are using OSRM in our system and this is a top needed feature for us to auto scale our platform. Is there a reason for keeping this PR still open? @gardster It would be great if you could resolve the conflicts.  
After conflict resolution, I also wonder if there will be any other obstacle left for merging the PR. Any thoughts? @akashihi

**Comment by akashihi** [comment]
@ugurakkurt definitely no obstacles

@gardster May i ask you to take a look on it?

**Comment by srimaln91** [comment]
Hi,

Any update on this? We use OSRM in our production systems and we would like to see this feature be released.

**Comment by srimaln91** [comment]
@gardster This is something that we would like to see in osrm-routed. May I ask you to take a look and resolve merge conflicts?

**Comment by DennisOSRM** [comment]
closing stale PR. Reopen if still relevant.

**Pull request closed without merging**